internal/cli/clients/push.go

```diff
@@ -1,96 +1,240 @@
+	Services     []string
+	Docker       string
+	GitRev       string
+	BuilderImage string
+	Name         string
+	Path         string
+}
+
+// Push pushes an app
+// * validate
+// * upload
+// * stage
+// * (tail logs)
+// * wait for pipelinerun
+// * deploy
+// * wait for app
+func (c *EpinioClient) Push(ctx context.Context, params PushParams) error {
+	name := params.Name
+	source := params.Path
+
+	appRef := models.AppRef{Name: name, Org: c.Config.Org}
+	log := c.Log.
+		WithName("Push").
+		WithValues("Name", appRef.Name,
+			"Namespace", appRef.Org,
+			"Sources", source)
+	log.Info("start")
+	defer log.Info("return")
+	details := log.V(1) // NOTE: Increment of level, not absolute. Visible via TRACE_LEVEL=2
 
-	// returns the source blob's UUID
-	upload := &models.UploadResponse{}
-	if err := json.Unmarshal(b, upload); err != nil {
-		return nil, err
+	sourceToShow := source
+	if params.GitRev != "" {
+		sourceToShow = fmt.Sprintf("%s @ %s", sourceToShow, params.GitRev)
 	}
 
-	return upload, nil
-}
+	msg := c.ui.Note().
+		WithStringValue("Name", appRef.Name).
+		WithStringValue("Sources", sourceToShow).
+		WithStringValue("Namespace", appRef.Org)
 
-// importGit asks the server to import a git repo and put in into the blob store
-func (c *EpinioClient) importGit(app models.AppRef, gitRef models.GitRef) (*models.ImportGitResponse, error) {
-	data := url.Values{}
-	data.Set("giturl", gitRef.URL)
-	data.Set("gitrev", gitRef.Revision)
+	if err := c.TargetOk(); err != nil {
+		return err
+	}
 
-	url := fmt.Sprintf("%s/%s", c.serverURL, api.Routes.Path("AppImportGit", app.Org, app.Name))
-	request, err := http.NewRequest("POST", url, strings.NewReader(data.Encode()))
-	if err != nil {
-		return nil, errors.Wrap(err, "constructing the request")
+	services := uniqueStrings(params.Services)
+
+	if len(services) > 0 {
+		sort.Strings(services)
+		msg = msg.WithStringValue("Services:", strings.Join(services, ", "))
 	}
-	request.SetBasicAuth(c.Config.User, c.Config.Password)
-	request.Header.Add("Content-Type", "application/x-www-form-urlencoded")
-	request.Header.Add("Content-Length", strconv.Itoa(len(data.Encode())))
 
-	response, err := (&http.Client{}).Do(request)
-	if err != nil {
-		return nil, errors.Wrap(err, "making the request to import git")
+	msg.Msg("About to push an application with given name and sources into the specified namespace")
+
+	c.ui.Exclamation().
+		Timeout(duration.UserAbort()).
+		Msg("Hit Enter to continue or Ctrl+C to abort (deployment will continue automatically in 5 seconds)")
+
+	details.Info("validate app name")
+	errorMsgs := validation.IsDNS1123Subdomain(appRef.Name)
+	if len(errorMsgs) > 0 {
+		return fmt.Errorf("%s: %s", "app name incorrect", strings.Join(errorMsgs, "\n"))
 	}
 
-	defer response.Body.Close()
-	bodyBytes, err := ioutil.ReadAll(response.Body)
+	c.ui.Normal().Msg("Create the application resource ...")
+
+	request := models.ApplicationCreateRequest{Name: appRef.Name}
+
+	_, err := c.API.AppCreate(
```

Comment:
Why does this method have an error func while others (e.g. `AppUpload` below) don't? Shouldn't we be consistent?

Reply:
Yes, I think there is something wrong with our http client helpers, they encapsulate the response struct. So the only way to see the real status code is to pass in a "custom error handler". Some API endpoints return a status code that's different from 200 and need to do this, because everything != 200 is considered an error by our helpers. 

I want to change that, maybe by exposing the response struct, but not in this PR.
Not a 100% sure how the result will look like.